File: internal/command/claims.go
package command

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/adamavenir/mini-msg/internal/db"
	"github.com/adamavenir/mini-msg/internal/types"
	"github.com/spf13/cobra"
)

// NewClaimsCmd creates the claims command.
func NewClaimsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "claims [agent]",
		Short: "List active claims",
		Args:  cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			claimType, _ := cmd.Flags().GetString("type")

			if _, err := db.PruneExpiredClaims(ctx.DB); err != nil {
				return writeCommandError(cmd, err)
			}

			var claims []types.Claim
			if len(args) > 0 {
				agentID := ResolveAgentRef(args[0], ctx.ProjectConfig)
				claims, err = db.GetClaimsByAgent(ctx.DB, agentID)
			} else {
				claims, err = db.GetAllClaims(ctx.DB)
			}
			if err != nil {
				return writeCommandError(cmd, err)
			}

			if claimType != "" {
				filtered := make([]types.Claim, 0, len(claims))
				for _, claim := range claims {
					if string(claim.ClaimType) == claimType {
						filtered = append(filtered, claim)
					}
				}
				claims = filtered
			}

			if ctx.JSONMode {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(claims)
			}

			out := cmd.OutOrStdout()
			if len(claims) == 0 {
				if len(args) > 0 {
					agentID := ResolveAgentRef(args[0], ctx.ProjectConfig)
					fmt.Fprintf(out, "No claims for @%s\n", agentID)
				} else {
					fmt.Fprintln(out, "No active claims")
				}
				return nil
			}

			fmt.Fprintf(out, "CLAIMS (%d):\n", len(claims))

			byAgent := map[string][]types.Claim{}
			for _, claim := range claims {
				byAgent[claim.AgentID] = append(byAgent[claim.AgentID], claim)
			}

			for agentID, agentClaims := range byAgent {
				fmt.Fprintf(out, "\n  @%s:\n", agentID)
				for _, claim := range agentClaims {
					typePrefix := ""
					if claim.ClaimType != types.ClaimTypeFile {
						typePrefix = fmt.Sprintf("%s:", claim.ClaimType)
					}
					age := formatRelative(claim.CreatedAt)
					expiry := ""
					if claim.ExpiresAt != nil {
						remaining := *claim.ExpiresAt - time.Now().Unix()
						if remaining > 0 {
							expiry = fmt.Sprintf(" (%dm left)", remaining/60)
						} else {
							expiry = " (expired)"
						}
					}
					reason := ""
					if claim.Reason != nil && *claim.Reason != "" {
						reason = " - " + *claim.Reason
					}
					fmt.Fprintf(out, "    %s%s (%s ago)%s%s\n", typePrefix, claim.Pattern, age, expiry, reason)
				}
			}

			return nil
		},
	}

	cmd.Flags().String("type", "", "filter by claim type (file, bd, issue)")
	return cmd
}

File: src/commands/claims.ts
import { Command } from 'commander';
import { getContext, handleError } from './shared.js';
import { getAllClaims, getClaimsByAgent } from '../db/queries.js';
import { resolveAgentRef } from '../core/context.js';
import { formatRelative } from '../core/time.js';
import type { ClaimType } from '../types.js';

export function claimsCommand(): Command {
  return new Command('claims')
    .description('List active claims')
    .argument('[agent]', 'filter by agent name (e.g., @alice)')
    .option('--type <type>', 'filter by claim type (file, bd, issue)')
    .action((agentArg: string | undefined, options, cmd) => {
      try {
        const { db, jsonMode, projectConfig } = getContext(cmd);

        let claims;
        if (agentArg) {
          const agentId = resolveAgentRef(agentArg, projectConfig ?? null);
          claims = getClaimsByAgent(db, agentId);
        } else {
          claims = getAllClaims(db);
        }

        // Filter by type if specified
        if (options.type) {
          const claimType = options.type as ClaimType;
          claims = claims.filter(c => c.claim_type === claimType);
        }

        if (jsonMode) {
          console.log(JSON.stringify(claims));
        } else {
          if (claims.length === 0) {
            if (agentArg) {
              const agentId = resolveAgentRef(agentArg, projectConfig ?? null);
              console.log(`No claims for @${agentId}`);
            } else {
              console.log('No active claims');
            }
          } else {
            console.log(`CLAIMS (${claims.length}):`);

            // Group by agent
            const byAgent = new Map<string, typeof claims>();
            for (const claim of claims) {
              const existing = byAgent.get(claim.agent_id) || [];
              existing.push(claim);
              byAgent.set(claim.agent_id, existing);
            }

            for (const [agentId, agentClaims] of byAgent) {
              console.log(`\n  @${agentId}:`);
              for (const claim of agentClaims) {
                const typePrefix = claim.claim_type === 'file' ? '' : `${claim.claim_type}:`;
                const age = formatRelative(claim.created_at);
                let expiry = '';
                if (claim.expires_at) {
                  const now = Math.floor(Date.now() / 1000);
                  if (claim.expires_at > now) {
                    const remaining = Math.round((claim.expires_at - now) / 60);
                    expiry = ` (${remaining}m left)`;
                  } else {
                    expiry = ' (expired)';
                  }
                }
                const reason = claim.reason ? ` - ${claim.reason}` : '';
                console.log(`    ${typePrefix}${claim.pattern} (${age})${expiry}${reason}`);
              }
            }
          }
        }

        db.close();
      } catch (error) {
        handleError(error);
      }
    });
}
